ResourceWarning from pytest with responses 0.24.0

### Describe the bug

Consider the following pytest test file:

```python
import requests
import responses

@responses.activate
def test_function():
    responses.get("http://www.example.com", body="Hello!")
    with requests.Session() as s:
        r = s.get("http://www.example.com")
        r.raise_for_status()
        assert r.text == "Hello!"
```

This is accompanied by the following pytest configuration in `tox.ini`:

```ini
[pytest]
filterwarnings = error  # Highly recommended in general
```

Running `pytest` on the above test succeeds without a problem with responses 0.23.3, but if responses 0.24.0 is used, the test fails with:

```
============================= test session starts ==============================
platform darwin -- Python 3.11.6, pytest-7.4.3, pluggy-1.3.0
rootdir: /Users/jwodder/work/dev/tmp/responses-bug
configfile: tox.ini
collected 1 item

test_case.py F                                                           [100%]

=================================== FAILURES ===================================
________________________________ test_function _________________________________

cls = <class '_pytest.runner.CallInfo'>
func = <function call_runtest_hook.<locals>.<lambda> at 0x1100e80e0>
when = 'call'
reraise = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)

    @classmethod
    def from_call(
        cls,
        func: "Callable[[], TResult]",
        when: "Literal['collect', 'setup', 'call', 'teardown']",
        reraise: Optional[
            Union[Type[BaseException], Tuple[Type[BaseException], ...]]
        ] = None,
    ) -> "CallInfo[TResult]":
        """Call func, wrapping the result in a CallInfo.
    
        :param func:
            The function to call. Called without arguments.
        :param when:
            The phase in which the function is called.
        :param reraise:
            Exception or exceptions that shall propagate if raised by the
            function, instead of being wrapped in the CallInfo.
        """
        excinfo = None
        start = timing.time()
        precise_start = timing.perf_counter()
        try:
>           result: Optional[TResult] = func()

.nox/new_responses/lib/python3.11/site-packages/_pytest/runner.py:341: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.nox/new_responses/lib/python3.11/site-packages/_pytest/runner.py:262: in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
.nox/new_responses/lib/python3.11/site-packages/pluggy/_hooks.py:493: in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
.nox/new_responses/lib/python3.11/site-packages/pluggy/_manager.py:115: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
.nox/new_responses/lib/python3.11/site-packages/_pytest/unraisableexception.py:88: in pytest_runtest_call
    yield from unraisable_exception_runtest_hook()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def unraisable_exception_runtest_hook() -> Generator[None, None, None]:
        with catch_unraisable_exception() as cm:
            yield
            if cm.unraisable:
                if cm.unraisable.err_msg is not None:
                    err_msg = cm.unraisable.err_msg
                else:
                    err_msg = "Exception ignored in"
                msg = f"{err_msg}: {cm.unraisable.object!r}\n\n"
                msg += "".join(
                    traceback.format_exception(
                        cm.unraisable.exc_type,
                        cm.unraisable.exc_value,
                        cm.unraisable.exc_traceback,
                    )
                )
>               warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
E               pytest.PytestUnraisableExceptionWarning: Exception ignored in: <socket.socket fd=-1, family=2, type=1, proto=0>
E               
E               Traceback (most recent call last):
E                 File "/usr/local/Cellar/python@3.11/3.11.6_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/socket.py", line 777, in close
E                   self._sock = None
E                   ^^^^^^^^^^
E               ResourceWarning: unclosed <socket.socket fd=14, family=2, type=1, proto=0, laddr=('0.0.0.0', 0)>

.nox/new_responses/lib/python3.11/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning
=========================== short test summary info ============================
FAILED test_case.py::test_function - pytest.PytestUnraisableExceptionWarning:...
============================== 1 failed in 0.35s ===============================
```

I suspect that the newer `responses` isn't closing a socket somewhere, possibly due to failure to use a context manager.

### Additional context

_No response_

### Version of `responses`

0.24.0

### Steps to Reproduce

I just filled that out above.

### Expected Result

No errors.

### Actual Result

See errors above.

--- a/CHANGES
+++ b/CHANGES
@@ -3,6 +3,7 @@
 
 * Reintroduced overloads for better typing in `CallList`.
 * Added typing to `Call` attributes.
+* Fix socket issues (see #693)
 
 
 0.24.0

--- a/responses/__init__.py
+++ b/responses/__init__.py
@@ -1,8 +1,6 @@
-import http
 import inspect
 import json as json_module
 import logging
-import socket
 from functools import partialmethod
 from functools import wraps
 from http import client
@@ -532,25 +530,38 @@ def _form_response(
     headers: Optional[Mapping[str, str]],
     status: int,
 ) -> HTTPResponse:
-    dummy_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-    orig_response = http.client.HTTPResponse(sock=dummy_socket)
     """
+    Function to generate `urllib3.response.HTTPResponse` object.
+
     The cookie handling functionality of the `requests` library relies on the response object
     having an original response object with the headers stored in the `msg` attribute.
     Instead of supplying a file-like object of type `HTTPMessage` for the headers, we provide
     the headers directly. This approach eliminates the need to parse the headers into a file-like
     object and then rely on the library to unparse it back. These additional conversions can
     introduce potential errors.
-    Therefore, we intentionally ignore type checking for this assignment.
     """
-    orig_response.msg = headers  # type: ignore[assignment]
 
+    data = BytesIO()
+    data.close()
+
+    """
+    The type `urllib3.response.HTTPResponse` is incorrect; we should
+    use `http.client.HTTPResponse` instead. However, changing this requires opening
+    a real socket to imitate the object. This may not be desired, as some users may
+    want to completely restrict network access in their tests.
+    See https://github.com/getsentry/responses/issues/691
+    """
+    orig_response = HTTPResponse(
+        body=data,  # required to avoid "ValueError: Unable to determine whether fp is closed."
+        msg=headers,  # type: ignore[arg-type]
+        preload_content=False,
+    )
     return HTTPResponse(
         status=status,
         reason=client.responses.get(status, None),
         body=body,
         headers=headers,
-        original_response=orig_response,
+        original_response=orig_response,  # type: ignore[arg-type]  # See comment above
         preload_content=False,
     )
 

--- a/responses/tests/test_responses.py
+++ b/responses/tests/test_responses.py
@@ -2054,7 +2054,7 @@ def test_response_calls_indexing_and_slicing():
         individual_call: Call = responses.calls[0]
         call_slice: List[Call] = responses.calls[1:-1]
 
-        assert individual_call.request.url == "http://www.example.com"
+        assert individual_call.request.url == "http://www.example.com/"
 
         assert call_slice == [
             responses.calls[1],

--- a/responses/tests/test_responses.py
+++ b/responses/tests/test_responses.py
@@ -2067,6 +2067,9 @@ def test_response_calls_indexing_and_slicing():
             "http://www.example.com/1",
         ]
 
+    run()
+    assert_reset()
+
 
 def test_response_calls_and_registry_calls_are_equal():
     @responses.activate

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,11 @@
 [tox]
 envlist = py38,py39,py310,py311,mypy,precom
 
+[pytest]
+filterwarnings =
+    error
+    default::DeprecationWarning
+
 [testenv]
 extras = tests
 commands =

